src/common/threads_manager.h

```diff
@@ -0,0 +1,157 @@
+ * <!> Be carefull in case of external container is copying or deleting.
+ */
+template <class DataT>
+class CyclicBuffer : public AbstractBuffer<DataT> {
+ public:
+  explicit CyclicBuffer(std::vector<DataT>* storage_ptr) : AbstractBuffer<DataT>(storage_ptr) {}
+
+  size_t Id() const override {
+    return this->counter_ % this->storage_ptr_->size();
+  }
+
+  size_t NCycles() const {
+    return this->counter_ / this->storage_ptr_->size();
+  }
+};
+
+/*
+ * This class provide an access with saturation to the data placed in external linear container of type std::vector<DataT>.
+ * Access with saturation means that in case of counter is higher than size - 1, than the last element is returning.
+ * <!> Be carefull in case of external container is copying or deleting.
+ */
+template <class DataT>
+class SaturationBuffer : public AbstractBuffer<DataT> {
+ public:
+  explicit SaturationBuffer(std::vector<DataT>* storage_ptr) : AbstractBuffer<DataT>(storage_ptr) {}
+
+  size_t Id() const override {
+    return this->counter_ < this->storage_ptr_->size() - 1
+           ? this->counter_
+           : this->storage_ptr_->size() - 1;
+  }
+};
+
+class ThreadsManager {
+ public:
+  struct ThreadInfo {
+    std::vector<Slice> addr;
+    std::vector<uint16_t> nodes_id;
+    std::unordered_map<uint32_t, uint32_t> nodes_count;
+    std::vector<uint32_t> rows_nodes_wise;
+
+    struct NodesCountRange {
+      uint32_t begin;
+      uint32_t end;
+    };
+
+    std::unordered_map<uint32_t, NodesCountRange> nodes_count_range;
+    NodesCountRange* GetNodesCountRangePtr(size_t nid) {
+     return &(nodes_count_range[nid]);
+    }
+
+    std::vector<uint32_t> vec_rows;
+    std::vector<uint32_t> vec_rows_remain;
+
+    std::unordered_map<uint32_t, size_t> states;
+    std::unordered_map<uint32_t, bool> default_flags;
+  };
+
+  struct NodeInfo {
+    std::vector<uint16_t> threads_id;
+  };
+
+  ThreadsManager() : threads(), nodes(), threads_cbuffer(&threads), threads_sbuffer(&threads) {}
+
+  ThreadsManager(const ThreadsManager& other) : threads(other.threads), nodes(other.nodes),
+                                                threads_cbuffer(&threads),
+                                                threads_sbuffer(&threads) {}
+
+  /* Tie thread_id with node_id */
+  void Tie(size_t tid, uint32_t nid) {
+    CHECK_LT(tid, threads.size());
+    if (nodes[nid].threads_id.empty()) {
+      nodes[nid].threads_id.push_back(tid);
+      threads[tid].nodes_id.push_back(nid);
+    } else {
+      if (nodes[nid].threads_id.back() != tid) {
+        nodes[nid].threads_id.push_back(tid);
+        threads[tid].nodes_id.push_back(nid);
+      }
+    }
+  }
+
+  void Init(size_t n_threads) {
+    threads.resize(n_threads);
+    nodes.clear();
+  }
+
+  ThreadInfo* GetThreadInfoPtr(size_t tid) {
+    return &(threads[tid]);
+  }
+
+  std::vector<ThreadInfo> threads;
+  std::unordered_map<uint32_t, NodeInfo> nodes;
+
+  CyclicBuffer<ThreadInfo> threads_cbuffer;
```

Comment:
make it private and add corresponding methods like GetThreadsCyclicView() and GetThreadsSaturatedView()